fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -0,0 +1,1839 @@
+	void returnIterator(ReadIterator& iter) {
+		if (SERVER_KNOBS->ROCKSDB_READ_RANGE_REUSE_ITERATORS) {
+			std::lock_guard<std::mutex> lock(mutex);
+			it = iteratorsMap.find(iter.index);
+			// iterator found: put the iterator back to the pool(inUse=false).
+			// iterator not found: update would have removed the iterator from pool, so nothing to do.
+			if (it != iteratorsMap.end()) {
+				ASSERT(it->second.inUse);
+				it->second.inUse = false;
+			}
+		}
+	}
+
+	// Called for every ROCKSDB_READ_RANGE_ITERATOR_REFRESH_TIME seconds in a loop.
+	void refreshIterators() {
+		std::lock_guard<std::mutex> lock(mutex);
+		it = iteratorsMap.begin();
+		while (it != iteratorsMap.end()) {
+			if (now() - it->second.creationTime > SERVER_KNOBS->ROCKSDB_READ_RANGE_ITERATOR_REFRESH_TIME) {
+				it = iteratorsMap.erase(it);
+			} else {
+				it++;
+			}
+		}
+	}
+
+	uint64_t numReadIteratorsCreated() { return index; }
+
+	uint64_t numTimesReadIteratorsReused() { return iteratorsReuseCount; }
+
+private:
+	std::unordered_map<int, ReadIterator> iteratorsMap;
+	std::unordered_map<int, ReadIterator>::iterator it;
+	DB& db;
+	rocksdb::ColumnFamilyHandle* cf;
+	rocksdb::ReadOptions readRangeOptions;
+	std::mutex mutex;
+	// incrementing counter for every new iterator creation, to uniquely identify the iterator in returnIterator().
+	uint64_t index;
+	uint64_t iteratorsReuseCount;
+};
+
+ACTOR Future<Void> flowLockLogger(const FlowLock* readLock, const FlowLock* fetchLock) {
+	loop {
+		wait(delay(SERVER_KNOBS->ROCKSDB_METRICS_DELAY));
+		TraceEvent e("RocksDBFlowLock");
+		e.detail("ReadAvailable", readLock->available());
+		e.detail("ReadActivePermits", readLock->activePermits());
+		e.detail("ReadWaiters", readLock->waiters());
+		e.detail("FetchAvailable", fetchLock->available());
+		e.detail("FetchActivePermits", fetchLock->activePermits());
+		e.detail("FetchWaiters", fetchLock->waiters());
+	}
+}
+
+struct PhysicalShard;
+
+struct DataShard {
+	DataShard(KeyRange range, PhysicalShard* physicalShard) : range(range), physicalShard(physicalShard) {}
+
+	KeyRange range;
+	PhysicalShard* physicalShard;
+};
+
+struct PhysicalShard {
+	PhysicalShard(rocksdb::DB* db, std::string id) : db(db), id(id) {}
+
+	void init() {
+		if (cf) {
+			return;
+		}
+		cf = createNewColumnFamily(db, id);
+		readIterPool = std::make_shared<ReadIteratorPool>(db, cf, id);
+	}
+
+	~PhysicalShard() {
+		readIterPool.reset();
+		if (!deletePending)
+			return;
+
+		// Destroy CF
+		auto s = db->DropColumnFamily(cf);
+		if (!s.ok()) {
+			logRocksDBError(s, "DestroyShard");
+			logShardEvent(id, ShardOp::DESTROY, SevError, s.ToString());
+			return;
+		}
+		logShardEvent(id, ShardOp::DESTROY);
+	}
+
+	rocksdb::DB* db;
+	std::string id;
+	rocksdb::ColumnFamilyHandle* cf;
+	std::unordered_map<std::string, std::unique_ptr<DataShard>> dataShards;
+	std::shared_ptr<ReadIteratorPool> readIterPool;
+	bool deletePending = false;
+};
+
+class ShardManager {
```

Comment:
This is a relatively thin layer? Why is this `ShardManager` necessary? Is it intended as a common interface between the KV store and underlying shards?

nit: add comments for this class.